R-package/tests/testthat/test_basic.R

```diff
@@ -684,6 +701,23 @@ test_that("lgb.train() respects parameter aliases for objective", {
   expect_equal(bst$params[["objective"]], "binary")
 })
 
+test_that("lgb.train() prefers objective in params to keyword argument", {
+  data("EuStockMarkets")
+  bst <- lgb.train(
+    data = lgb.Dataset(
+      data = EuStockMarkets[, c("SMI", "CAC", "FTSE")]
+      , label = EuStockMarkets[, "DAX"]
+    )
+    , params = list(
+        objective = "regression_l2"
+        , verbosity = -1L
+    )
+    , nrounds = 5L
+    , obj = "regression_l1"
+  )
+  expect_equal(bst$params$objective, "regression_l2")
```

Comment:
TBH, it doesn't look like a reliable test. Simple preserving params object "as-is" as a field of a Booster will pass this check. Maybe we can at least use some non-trivial `objective` param alias in `params`?

Reply:
I agree with you. I just pushed https://github.com/microsoft/LightGBM/pull/5007/commits/297a0a561c37f4facbcf7c77b0328c4fc096d40c to make this a lot stricter.

* test that `objective` in the model dumped by by `$save_model_to_string()` matches the one from `params`
    - a lot stricter, since it relies on a code path that will extract the objective from the `Booster` object on the C++ side, instead of just the `params` list in the R object
    - https://github.com/microsoft/LightGBM/blob/9fc348af6f5ca60b91c9447d813a401567a4738a/R-package/src/lightgbm_R.cpp#L867
* use aliases for `objective` in params